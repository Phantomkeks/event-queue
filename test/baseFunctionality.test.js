"use strict";

const path = require("path");

const cds = require("@sap/cds/lib");

const eventQueue = require("../src");
const testHelper = require("./helper");
const EventQueueTest = require("./asset/EventQueueTest");
jest.mock("../src/shared/logger", () => require("./mocks/logger"));
const loggerMock = require("../src/shared/logger").Logger();

const project = __dirname + "/.."; // The project's root folder
cds.test(project);

describe("baseFunctionality", () => {
  let context, tx;

  beforeAll(async () => {
    const configFilePath = path.join(__dirname, "asset", "config.yml");
    await eventQueue.initialize({ configFilePath, registerDbHandler: false });
  });

  beforeEach(async () => {
    context = new cds.EventContext({ user: "testUser", tenant: 123 });
    tx = cds.tx(context);
    await tx.run(DELETE.from("sap.core.EventLock"));
    await tx.run(DELETE.from("sap.core.EventQueue"));
  });

  afterEach(async () => {
    await tx.rollback();
    jest.clearAllMocks();
  });

  afterAll(() => cds.shutdown);

  test("empty queue - nothing to do", async () => {
    const event = eventQueue.getConfigInstance().events[0];
    await eventQueue.processEventQueue(context, event.type, event.subType);
    expect(loggerMock.callsLengths().error).toEqual(0);
  });

  test("insert one entry and process", async () => {
    await testHelper.insertEventEntry(tx);
    const event = eventQueue.getConfigInstance().events[0];
    await eventQueue.processEventQueue(context, event.type, event.subType);
    expect(loggerMock.callsLengths().error).toEqual(0);
    await testHelper.selectEventQueueAndExpectDone(tx);
  });

  test("should do nothing if no lock is available", async () => {
    await testHelper.insertEventEntry(tx);
    jest
      .spyOn(
        eventQueue.EventQueueProcessorBase.prototype,
        "handleDistributedLock"
      )
      .mockResolvedValueOnce(false);
    const event = eventQueue.getConfigInstance().events[0];
    await eventQueue.processEventQueue(context, event.type, event.subType);
    expect(loggerMock.callsLengths().error).toEqual(0);
    await testHelper.selectEventQueueAndExpectOpen(tx);
  });

  describe("error handling", () => {
    test("missing event implementation", async () => {
      await testHelper.insertEventEntry(tx);
      await eventQueue.processEventQueue(context, "404", "NOT FOUND");
      expect(loggerMock.callsLengths().error).toEqual(1);
      expect(loggerMock.calls().error).toMatchSnapshot();
      await testHelper.selectEventQueueAndExpectOpen(tx);
    });

    test("handle handleDistributedLock fails", async () => {
      await testHelper.insertEventEntry(tx);
      jest
        .spyOn(
          eventQueue.EventQueueProcessorBase.prototype,
          "handleDistributedLock"
        )
        .mockRejectedValueOnce(new Error("lock require failed"));
      const event = eventQueue.getConfigInstance().events[0];
      await eventQueue.processEventQueue(context, event.type, event.subType);
      expect(loggerMock.callsLengths().error).toEqual(1);
      expect(loggerMock.calls().error).toMatchSnapshot();
      await testHelper.selectEventQueueAndExpectOpen(tx);
    });

    test("handle getQueueEntriesAndSetToInProgress fails", async () => {
      await testHelper.insertEventEntry(tx);
      jest
        .spyOn(
          eventQueue.EventQueueProcessorBase.prototype,
          "getQueueEntriesAndSetToInProgress"
        )
        .mockRejectedValueOnce(new Error("db error"));
      const event = eventQueue.getConfigInstance().events[0];
      await eventQueue.processEventQueue(context, event.type, event.subType);
      expect(loggerMock.callsLengths().error).toEqual(1);
      expect(loggerMock.calls().error).toMatchSnapshot();
      await testHelper.selectEventQueueAndExpectOpen(tx);
    });

    test("handle modifyQueueEntry fails", async () => {
      await testHelper.insertEventEntry(tx);
      jest
        .spyOn(eventQueue.EventQueueProcessorBase.prototype, "modifyQueueEntry")
        .mockImplementationOnce(() => {
          throw new Error("syntax error");
        });
      const event = eventQueue.getConfigInstance().events[0];
      await eventQueue.processEventQueue(context, event.type, event.subType);
      expect(loggerMock.callsLengths().error).toEqual(1);
      expect(loggerMock.calls().error).toMatchSnapshot();
      await testHelper.selectEventQueueAndExpectError(tx);
    });

    test("handle checkEventAndGeneratePayload fails", async () => {
      await testHelper.insertEventEntry(tx);
      jest
        .spyOn(EventQueueTest.prototype, "checkEventAndGeneratePayload")
        .mockRejectedValueOnce(new Error("syntax error"));
      const event = eventQueue.getConfigInstance().events[0];
      await eventQueue.processEventQueue(context, event.type, event.subType);
      expect(loggerMock.callsLengths().error).toEqual(1);
      expect(loggerMock.calls().error).toMatchSnapshot();
      await testHelper.selectEventQueueAndExpectError(tx);
    });

    test("handle clusterQueueEntries fails", async () => {
      await testHelper.insertEventEntry(tx);
      jest
        .spyOn(
          eventQueue.EventQueueProcessorBase.prototype,
          "clusterQueueEntries"
        )
        .mockImplementationOnce(() => {
          throw new Error("syntax error");
        });
      const event = eventQueue.getConfigInstance().events[0];
      await eventQueue.processEventQueue(context, event.type, event.subType);
      expect(loggerMock.callsLengths().error).toEqual(1);
      expect(loggerMock.calls().error).toMatchSnapshot();
      await testHelper.selectEventQueueAndExpectError(tx);
    });
  });

  describe("insert events", () => {
    test("straight forward", async () => {
      const event = eventQueue.getConfigInstance().events[0];
      await eventQueue.publishEvent(tx, {
        type: event.type,
        subType: event.subType,
        payload: JSON.stringify({
          testPayload: 123,
        }),
      });
      const events = await tx.run(SELECT.from("sap.core.EventQueue"));
      expect(events).toHaveLength(1);
      expect(events[0]).toMatchObject({
        type: event.type,
        subType: event.subType,
        payload: JSON.stringify({
          testPayload: 123,
        }),
      });
    });

    test("unknown event", async () => {
      try {
        await eventQueue.publishEvent(tx, {
          type: "404",
          subType: "NOT FOUND",
          payload: JSON.stringify({
            testPayload: 123,
          }),
        });
      } catch (err) {
        expect(err.toString()).toMatchInlineSnapshot(`"EventQueueError"`);
      }
      const events = await tx.run(SELECT.from("sap.core.EventQueue"));
      expect(events).toHaveLength(0);
    });
  });
});
