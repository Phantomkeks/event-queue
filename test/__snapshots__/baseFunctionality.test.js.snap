// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseFunctionality error handling handle checkEventAndGeneratePayload fails 1`] = `
[
  [
    "Unexpected error during event processing - setting queue entry to error. Error: Error: syntax error",
    {
      "eventSubType": "Task",
      "eventType": "Notifications",
      "queueEntriesIds": [
        "dbaa22d5-41db-4ff3-bdd8-e0bb19b217cf",
      ],
    },
  ],
]
`;

exports[`baseFunctionality error handling handle clusterQueueEntries fails 1`] = `
[
  [
    "Error during clustering of events - setting all queue entries to error. Error: Error: syntax error",
    {
      "eventSubType": "Task",
      "eventType": "Notifications",
    },
  ],
]
`;

exports[`baseFunctionality error handling handle getQueueEntriesAndSetToInProgress fails 1`] = `
[
  [
    "Processing event queue failed with unexpected error",
    {
      "error": [ExecuteInNewTransactionError: Execution in new transaction failed: db error],
    },
  ],
]
`;

exports[`baseFunctionality error handling handle handleDistributedLock fails 1`] = `
[
  [
    "Processing event queue failed with unexpected error",
    {
      "error": [Error: lock require failed],
    },
  ],
]
`;

exports[`baseFunctionality error handling handle modifyQueueEntry fails 1`] = `
[
  [
    "Unexpected error during event processing - setting queue entry to error. Error: Error: syntax error",
    {
      "eventSubType": "Task",
      "eventType": "Notifications",
      "queueEntriesIds": [
        "dbaa22d5-41db-4ff3-bdd8-e0bb19b217cf",
      ],
    },
  ],
]
`;

exports[`baseFunctionality error handling missing event implementation 1`] = `
[
  [
    "No Implementation found for queue type in 'eventTypeRegister.js'",
    {
      "eventSubType": "NOT FOUND",
      "eventType": "404",
    },
  ],
]
`;
